File: pkm_rs/src/ohpkm/convert/pk9.rs
use pkm_rs_resources::ball::Ball;
use pkm_rs_resources::{metadata_source::MetadataSource, ribbons::OpenHomeRibbonSet};
use pkm_rs_types::FlagSet;
use pkm_rs_types::{AbilityNumber, Stats16Le};

use super::OhpkmConvert;
use crate::convert_strategy::ConvertStrategy;
use crate::convert_strategy::PkmConverter;
use crate::format::PkmFormat;
use crate::gen8_swsh;
use crate::gen9_sv::Pk9;
use crate::ohpkm;
use crate::ohpkm::OhpkmV2;
use crate::ohpkm::v2_sections::ScarletVioletData;
use crate::ohpkm::v2_sections::pkm_bytes::StoredPkmBytes;
use crate::result::{Error, Result};
use crate::traits::HasSpeciesAndForm;

impl OhpkmConvert for Pk9 {
    fn to_main_data(&self) -> ohpkm::v2_sections::MainDataV2 {
        ohpkm::v2_sections::MainDataV2 {
            personality_value: self.personality_value,
            encryption_constant: self.encryption_constant,
            species_and_form: self.species_and_form.into_inner(),
            held_item_index: self.held_item_index,
            trainer_id: self.trainer_id,
            secret_id: self.secret_id,
            exp: self.exp,
            ability_index: self
                .ability_index
                .change_bound()
                .expect("Pk9 max ability <= overall max ability"),
            ability_num: self.ability_num,
            markings: self.markings,
            nature: self.nature,
            mint_nature: if self.mint_nature != self.nature {
                Some(self.mint_nature)
            } else {
                None
            },
            is_fateful_encounter: self.is_fateful_encounter,
            gender: self.gender,
            evs: self.evs,
            contest: self.contest,
            pokerus_byte: Default::default(),
            contest_memory_count: self.contest_memory_count,
            battle_memory_count: self.battle_memory_count,
            sociability: Default::default(),
            height_scalar: self.height_scalar,
            weight_scalar: self.weight_scalar,
            ribbons: OpenHomeRibbonSet::from_modern(self.ribbons),
            moves: self
                .moves
                .to_pp_adjusted(MetadataSource::SwordShield, ohpkm::MOVE_METADATA_SOURCE),
            nickname: self.nickname,
            relearn_moves: self.relearn_moves,
            ivs: self.ivs,
            is_egg: self.is_egg,
            is_nicknamed: self.is_nicknamed,
            handler_name: self.handler_name,
            is_current_handler: self.is_current_handler,
            handler_friendship: self.handler_friendship,
            handler_memory: self.handler_memory,
            handler_gender: self.handler_gender,
            handler_language: self.handler_language,
            fullness: Default::default(),
            enjoyment: Default::default(),
            game_of_origin: self.game_of_origin,
            game_of_origin_battle: self.game_of_origin_battle,
            language: self.language,
            form_argument: self.form_argument,
            affixed_ribbon: self.affixed_ribbon,
            trainer_name: self.trainer_name,
            trainer_friendship: self.trainer_friendship,
            trainer_memory: self.trainer_memory,
            egg_date: self.egg_date,
            met_date: self.met_date,
            ball: self.ball,
            egg_location_index: if self.egg_location_index == 0 {
                None
            } else {
                Some(self.egg_location_index)
            },
            met_location_index: self.met_location_index,
            met_level: self.met_level,
            hyper_training: self.hyper_training,
            trainer_gender: self.trainer_gender,
            home_tracker: self.home_tracker,
            favorite: Default::default(),
            is_shadow: Default::default(),
            scale: self.scale,
            handler_id: Default::default(),
            handler_affection: Default::default(),
            console_region: Default::default(),
            extra_form: Default::default(),
            trainer_affection: Default::default(),
            obedience_level: self.obedience_level,
            display_color_rgb: Default::default(),
            started_tracking_seconds: Default::default(),
            pid_bit_flipped_for_shiny: Default::default(),
        }
    }

    fn to_sv_data(&self) -> Option<ScarletVioletData> {
        Some(ScarletVioletData {
            tera_type_original: self.tera_type_original,
            tera_type_override: self.tera_type_override,
            tm_flags: self.tm_flags_base_game,
            tm_flags_dlc: self.tm_flags_dlc,
        })
    }

    fn from_ohpkm(ohpkm: &OhpkmV2, strategy: ConvertStrategy) -> Result<Self> {
        let form_metadata = ohpkm.get_forme_metadata();
        let converter = PkmConverter::new(PkmFormat::PK9, strategy);
        let met_data = converter.met_data(ohpkm);

        let ability_index = if let Some(current_ability) = ohpkm.ability_index().change_bound() {
            current_ability
        } else if let Some(ability_from_num) = form_metadata
            .get_ability(ohpkm.ability_num())
            .change_bound()
        {
            ability_from_num
        } else if let Some(first_ability) = form_metadata
            .get_ability(AbilityNumber::First)
            .change_bound()
        {
            first_ability
        } else {
            return Err(Error::AbilityIndex {
                ability_index: form_metadata.get_ability(AbilityNumber::First).to_u16(),
            });
        };

        let mut mon =
            Self {
                encryption_constant: ohpkm.encryption_constant(),
                sanity: 0,
                checksum: 0,
                species_and_form: ohpkm.species_and_form().try_into()?,
                held_item_index: ohpkm.held_item_index(),
                trainer_id: ohpkm.trainer_id(),
                secret_id: ohpkm.secret_id(),
                exp: ohpkm.exp(),
                ability_index,
                ability_num: ohpkm.ability_num(),
                markings: ohpkm.markings(),
                personality_value: ohpkm.personality_value(),
                nature: ohpkm.nature(),
                mint_nature: ohpkm.mint_nature(),
                is_fateful_encounter: ohpkm.is_fateful_encounter(),
                gender: ohpkm.gender(),
                evs: ohpkm.evs(),
                contest: ohpkm.contest(),
                ribbons: ohpkm.ribbons().get_modern().into_iter().collect(),
                height_scalar: ohpkm.height_scalar(),
                weight_scalar: ohpkm.weight_scalar(),
                scale: ohpkm.scale(),
                contest_memory_count: ohpkm.contest_memory_count(),
                battle_memory_count: ohpkm.battle_memory_count(),
                form_argument: ohpkm.form_argument(),
                affixed_ribbon: ohpkm.affixed_ribbon(),
                nickname: ohpkm.nickname(),
                moves: ohpkm
                    .moves()
                    .to_pp_adjusted(ohpkm::MOVE_METADATA_SOURCE, MetadataSource::SwordShield),
                relearn_moves: ohpkm.relearn_moves(),
                ivs: converter.ivs(ohpkm),
                is_egg: ohpkm.is_egg(),
                is_nicknamed: ohpkm.is_nicknamed(),
                handler_name: ohpkm.handler_name(),
                handler_gender: ohpkm.handler_gender(),
                is_current_handler: ohpkm.is_current_handler(),
                handler_friendship: ohpkm.handler_friendship(),
                handler_memory: ohpkm.handler_memory(),
                handler_language: ohpkm.handler_language(),
                handler_id: ohpkm.handler_id(),
                trainer_name: ohpkm.trainer_name(),
                trainer_friendship: ohpkm.trainer_friendship(),
                trainer_memory: ohpkm.trainer_memory(),
                egg_date: ohpkm.egg_date(),
                met_date: ohpkm.met_date(),
                obedience_level: ohpkm.obedience_level(),
                egg_location_index: ohpkm.egg_location_index().unwrap_or(0),
                met_location_index: met_data.location_index,
                ball: if ohpkm.ball() <= Ball::Beast {
                    ohpkm.ball()
                } else {
                    Ball::Poke
                },
                met_level: ohpkm.met_level(),
                trainer_gender: ohpkm.trainer_gender(),
                tm_flags_base_game: ohpkm
                    .tm_flags_sv()
                    .map(|v| {
                        FlagSet::from_bytes(v.try_into().expect(
                        "Scarlet/Violet base game TM flags are the expected bytelength in OHPKM",
                    ))
                    })
                    .unwrap_or_default(),
                tm_flags_dlc: ohpkm
                    .tm_flags_sv_dlc()
                    .map(|v| {
                        FlagSet::from_bytes(v.try_into().expect(
                            "Scarlet/Violet DLC TM flags are the expected bytelength in OHPKM",
                        ))
                    })
                    .unwrap_or_default(),
                home_tracker: ohpkm.home_tracker(),
                hyper_training: ohpkm.hyper_training(),
                game_of_origin: met_data.origin,
                game_of_origin_battle: ohpkm.game_of_origin_battle(),
                language: ohpkm.language(),
                tera_type_original: ohpkm.tera_type_original(),
                tera_type_override: ohpkm.tera_type_override(),
                status_condition: 0,
                stat_level: 0,
                current_hp: 0,
                stats: Stats16Le::default(),
            };

        mon.stat_level = mon.calculate_level();
        mon.stats = mon.calculate_stats();
        mon.current_hp = mon.stats.hp;

        mon.refresh_checksum();

        Ok(mon)
    }

    fn bytes_to_stored(bytes: &[u8]) -> Result<StoredPkmBytes> {
        bytes
            .try_into()
            .map_err(|_| {
                Error::buffer_size_with_source(
                    "Pk9::OhpkmConvert::bytes_to_stored",
                    gen8_swsh::PKM_DATA_SIZE,
                    bytes.len(),
                )
            })
            .map(StoredPkmBytes::Pk9)
    }
}

File: src/core/pkm/PK9.ts
import { OHPKM } from '@openhome-core/pkm/OHPKM'
import { ModernRibbons } from '@openhome-core/resources'
import { Errorable, Option, R } from '@openhome-core/util/functional'
import { FourMoves, PKMDate, Stats } from '@openhome-core/util/types'
import {
  AbilityIndex,
  BinaryGender,
  ContestStats,
  ConvertStrategies,
  ConvertStrategy,
  HyperTraining,
  Item,
  Language,
  MarkingsSixShapesColors,
  MetadataSummaryLookup,
  ModernRibbon,
  NatureIndex,
  OriginGame,
  Pk9Wasm,
  PkmFormat,
  PokeDate,
  SpeciesLookup,
  TeraTypeWasm,
  TrainerMemory,
} from '@pkm-rs/pkg'
import { PkmConstructorOptions } from './PKM'
import * as encryption from './util/encryption'
import { convertPokeDate, convertPokeDateOptional } from './wasm/convert'

export default class PK9 {
  format: PkmFormat = 'PK9'

  inner: Pk9Wasm

  constructor(arg: OHPKM | Pk9Wasm, options: PkmConstructorOptions) {
    if (arg instanceof Pk9Wasm) {
      this.inner = arg
    } else {
      const ohpkmBytes = new Uint8Array(arg.toBytes())

      this.inner = Pk9Wasm.fromOhpkmBytes(
        ohpkmBytes,
        options.strategy || ConvertStrategies.getDefault()
      )
    }
  }

  static getFormat(): PkmFormat {
    return 'PK9'
  }

  static getBoxSize() {
    return 344
  }

  static fromBytes(buffer: ArrayBuffer, encrypted?: boolean): PK9 {
    const byteArray = new Uint8Array(buffer)
    return PK9.fromWasm(
      encrypted ? Pk9Wasm.fromEncryptedBytes(byteArray) : Pk9Wasm.fromBytes(byteArray)
    )
  }

  static fromOhpkm(ohpkm: OHPKM, strategy: ConvertStrategy): Errorable<PK9> {
    return R.tryFrom(() => new PK9(ohpkm, { strategy }))
  }

  static fromWasm(pk9: Pk9Wasm): PK9 {
    return new PK9(pk9, {})
  }

  get encryptionConstant() {
    return this.inner.encryption_constant
  }
  set encryptionConstant(value: number) {
    this.inner.encryption_constant = value
  }

  get sanity() {
    return this.inner.sanity
  }
  set sanity(value: number) {
    this.inner.sanity = value
  }

  get checksum() {
    return this.inner.checksum
  }
  set checksum(value: number) {
    this.inner.checksum = value
  }

  get dexNum() {
    return this.inner.nationalDex
  }

  get heldItemIndex() {
    return this.inner.held_item_index
  }
  set heldItemIndex(value: number) {
    this.inner.held_item_index = value
  }

  get trainerID() {
    return this.inner.trainer_id
  }
  set trainerID(value: number) {
    this.inner.trainer_id = value
  }

  get secretID() {
    return this.inner.secret_id
  }
  set secretID(value: number) {
    this.inner.secret_id = value
  }

  get exp() {
    return this.inner.exp
  }
  set exp(value: number) {
    this.inner.exp = value
  }

  get ability() {
    return this.inner.abilityIndex
  }
  set ability(value: AbilityIndex) {
    this.inner.abilityIndex = AbilityIndex.fromIndex(value.index) || value
  }

  get abilityNum() {
    return this.inner.ability_num
  }
  set abilityNum(value: number) {
    this.inner.ability_num = value
  }

  get markings() {
    return this.inner.markings
  }
  set markings(value: MarkingsSixShapesColors) {
    this.inner.markings = value
  }

  get personalityValue() {
    return this.inner.personality_value
  }
  set personalityValue(value: number) {
    this.inner.personality_value = value
  }

  get nature() {
    return this.inner.nature.copy()
  }
  set nature(value: NatureIndex) {
    this.inner.nature = value.copy()
  }

  get statNature() {
    return this.inner.mint_nature.copy()
  }
  set statNature(value: NatureIndex) {
    this.inner.mint_nature = value.copy()
  }

  get isFatefulEncounter() {
    return this.inner.is_fateful_encounter
  }
  set isFatefulEncounter(value: boolean) {
    this.inner.is_fateful_encounter = value
  }

  get gender() {
    return this.inner.gender
  }
  set gender(value: number) {
    this.inner.gender = value
  }

  get formNum() {
    return this.inner.formIndex
  }

  get evs() {
    return this.inner.evs
  }
  set evs(value: Stats) {
    this.inner.evs = value
  }

  get contest() {
    return this.inner.contest
  }
  set contest(value: ContestStats) {
    this.inner.contest = value
  }

  get contestMemoryCount() {
    return this.inner.contest_memory_count
  }
  set contestMemoryCount(value: number) {
    this.inner.contest_memory_count = value
  }

  get battleMemoryCount() {
    return this.inner.battle_memory_count
  }
  set battleMemoryCount(value: number) {
    this.inner.battle_memory_count = value
  }

  get heightScalar() {
    return this.inner.height_scalar
  }
  set heightScalar(value: number) {
    this.inner.height_scalar = value
  }

  get weightScalar() {
    return this.inner.weight_scalar
  }
  set weightScalar(value: number) {
    this.inner.weight_scalar = value
  }

  get scale() {
    return this.inner.scale
  }
  set scale(value: number) {
    this.inner.scale = value
  }

  get formArgument() {
    return this.inner.form_argument
  }
  set formArgument(value: number) {
    this.inner.form_argument = value
  }

  get nickname() {
    return this.inner.nickname
  }
  set nickname(value: string) {
    this.inner.nickname = value
  }

  get moves() {
    const moves = Array.from(this.inner.move_indices)
    if (moves.length !== 4) {
      throw new Error(`PK9 WASM struct has move array length of ${moves.length} (expected 4)`)
    }

    return moves as FourMoves
  }
  set moves(value: FourMoves) {
    this.inner.move_indices = new Uint16Array(value)
  }

  get movePP() {
    const movePP = Array.from(this.inner.move_pp)
    if (movePP.length !== 4) {
      throw new Error(`PK9 WASM struct has move PP array length of ${movePP.length} (expected 4)`)
    }

    return movePP as FourMoves
  }
  set movePP(value: FourMoves) {
    this.inner.move_pp = new Uint8Array(value)
  }

  get movePPUps() {
    const movePPUps = Array.from(this.inner.move_pp_ups)
    if (movePPUps.length !== 4) {
      throw new Error(
        `PK9 WASM struct has move PP up array length of ${movePPUps.length} (expected 4)`
      )
    }

    return movePPUps as FourMoves
  }
  set movePPUps(value: FourMoves) {
    this.inner.move_pp_ups = new Uint8Array(value)
  }

  get relearnMoves() {
    const relearnMoves = Array.from(this.inner.move_pp_ups)
    if (relearnMoves.length !== 4) {
      throw new Error(
        `PK9 WASM struct has relearn move array length of ${relearnMoves.length} (expected 4)`
      )
    }
    return relearnMoves as FourMoves
  }
  set relearnMoves(value: FourMoves) {
    this.inner.relearn_move_indices = new Uint16Array(value)
  }

  get ivs() {
    return this.inner.ivs
  }
  set ivs(value: Stats) {
    this.inner.ivs = value
  }

  get isEgg() {
    return this.inner.is_egg
  }
  set isEgg(value: boolean) {
    this.inner.is_egg = value
  }

  get isNicknamed() {
    return this.inner.is_nicknamed
  }
  set isNicknamed(value: boolean) {
    this.inner.is_nicknamed = value
  }

  get teraTypeOriginal() {
    return this.inner.teraTypeOriginal
  }
  set teraTypeOriginal(value: TeraTypeWasm) {
    this.inner.teraTypeOriginal = value
  }

  get teraTypeOverride() {
    return this.inner.teraTypeOverride
  }
  set teraTypeOverride(value: Option<TeraTypeWasm>) {
    this.inner.teraTypeOverride = value
  }

  get obedienceLevel() {
    return this.inner.obedienceLevel
  }
  set obedienceLevel(value: number) {
    this.inner.obedienceLevel = value
  }

  get handlerName() {
    return this.inner.handler_name
  }
  set handlerName(value: string) {
    this.inner.handler_name = value
  }

  get handlerGender() {
    return this.inner.handler_gender
  }
  set handlerGender(value: BinaryGender) {
    this.inner.handler_gender = value
  }

  get handlerLanguage() {
    return this.inner.handler_language
  }
  set handlerLanguage(value: Option<Language>) {
    this.inner.handler_language = value
  }

  get isCurrentHandler() {
    return this.inner.is_current_handler
  }
  set isCurrentHandler(value: boolean) {
    this.inner.is_current_handler = value
  }

  get handlerFriendship() {
    return this.inner.handler_friendship
  }
  set handlerFriendship(value: number) {
    this.inner.handler_friendship = value
  }

  get trainerName() {
    return this.inner.trainer_name
  }
  set trainerName(value: string) {
    this.inner.trainer_name = value
  }

  get trainerFriendship() {
    return this.inner.trainer_friendship
  }
  set trainerFriendship(value: number) {
    this.inner.trainer_friendship = value
  }

  get eggDate() {
    return convertPokeDateOptional(this.inner.egg_date)
  }

  set eggDate(value: PKMDate | undefined) {
    if (value) {
      this.inner.egg_date = new PokeDate(value.year, value.month, value.day)
    } else {
      this.inner.egg_date = undefined
    }
  }

  get metDate() {
    return convertPokeDate(this.inner.met_date)
  }
  set metDate(value: PKMDate) {
    this.inner.met_date = new PokeDate(value.year, value.month, value.day)
  }

  get eggLocationIndex() {
    return this.inner.egg_location_index
  }
  set eggLocationIndex(value: number) {
    this.inner.egg_location_index = value
  }

  get metLocationIndex() {
    return this.inner.met_location_index
  }
  set metLocationIndex(value: number) {
    this.inner.met_location_index = value
  }

  get ball() {
    return this.inner.ball
  }
  set ball(value: number) {
    this.inner.ball = value
  }

  get metLevel() {
    return this.inner.met_level
  }
  set metLevel(value: number) {
    this.inner.met_level = value
  }

  get hyperTraining() {
    return this.inner.hyper_training
  }
  set hyperTraining(value: HyperTraining) {
    this.inner.hyper_training = value
  }

  get gameOfOrigin() {
    return this.inner.game_of_origin
  }
  set gameOfOrigin(value: OriginGame) {
    this.inner.game_of_origin = value
  }

  get gameOfOriginBattle() {
    return this.inner.game_of_origin_battle
  }
  set gameOfOriginBattle(value: Option<OriginGame>) {
    this.inner.game_of_origin_battle = value
  }

  get language() {
    return this.inner.language
  }
  set language(value: number) {
    this.inner.language = value
  }

  get statusCondition() {
    return this.inner.status_condition
  }
  set statusCondition(value: number) {
    this.inner.status_condition = value
  }

  get currentHP() {
    return this.inner.current_hp
  }
  set currentHP(value: number) {
    this.inner.current_hp = value
  }

  get ribbons() {
    return this.inner.ribbons.map((ribbonName) =>
      ribbonName.endsWith('Ribbon') ? ribbonName.substring(0, ribbonName.length - 7) : ribbonName
    )
  }
  set ribbons(ribbonNames: string[]) {
    this.inner.ribbon_indices = new Uint32Array(
      ribbonNames.map((name) => ModernRibbons.indexOf(name)).filter((idx) => idx !== -1)
    )
  }

  get affixedRibbon() {
    return this.inner.affixed_ribbon
  }
  set affixedRibbon(value: Option<ModernRibbon>) {
    this.inner.affixed_ribbon = value
  }

  get handlerMemory() {
    return this.inner.handler_memory
  }
  set handlerMemory(value: TrainerMemory) {
    this.inner.handler_memory = value
  }

  get trainerMemory() {
    return this.inner.trainer_memory
  }
  set trainerMemory(value: TrainerMemory) {
    this.inner.trainer_memory = value
  }

  get trainerGender() {
    return this.inner.trainer_gender
  }
  set trainerGender(value: BinaryGender) {
    this.inner.trainer_gender = value
  }

  get tmFlagsBaseGame() {
    return this.inner.tmFlagsBaseGame
  }

  get tmFlagsDlc() {
    return this.inner.tmFlagsDlc
  }

  get homeTracker() {
    return this.inner.home_tracker
  }
  set homeTracker(value: Option<bigint>) {
    this.inner.home_tracker = value
  }

  getStats() {
    return this.inner.calculateStats()
  }

  // stored stats
  get stats() {
    return this.inner.stats
  }

  get abilityName() {
    return this.inner.abilityIndex.name
  }

  toBytes() {
    return this.inner.toBytes().buffer as ArrayBuffer
  }

  get heldItemName() {
    return Item.fromIndex(this.heldItemIndex)?.name ?? 'None'
  }

  calculateChecksum() {
    return encryption.get16BitChecksumLittleEndian(this.toBytes(), 0x08, 0x148)
  }

  refreshChecksum() {
    this.checksum = encryption.get16BitChecksumLittleEndian(this.toBytes(), 0x08, 0x148)
  }

  toPCBytes() {
    const shuffledBytes = encryption.shuffleBlocksGen89(this.toBytes())
    return encryption.decryptByteArrayGen89(shuffledBytes)
  }

  getLevel() {
    return this.speciesMetadata?.calculateLevel(this.exp) ?? 1
  }

  isShiny() {
    return (
      (this.trainerID ^
        this.secretID ^
        (this.personalityValue & 0xffff) ^
        ((this.personalityValue >> 16) & 0xffff)) <
      16
    )
  }

  isSquareShiny() {
    return !(
      this.trainerID ^
      this.secretID ^
      (this.personalityValue & 0xffff) ^
      ((this.personalityValue >> 16) & 0xffff)
    )
  }

  public get metadata() {
    return MetadataSummaryLookup(this.dexNum, this.formNum)
  }

  public get speciesMetadata() {
    return SpeciesLookup(this.dexNum)
  }

  static maxValidMove() {
    return 919
  }

  static maxValidBall() {
    return 0
  }

  static allowedBalls() {
    return [
      1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
      27,
    ]
  }
}
